File: py-files/utils_translated.py

from typing import Iterator, TextIO, ItemsView, Callable, Union
from abc import ABC
from dataclasses import dataclass
from ipaddress import IPv4Address
from collections import deque
from functools import partial
from builtins_translated import Operator

type op_result_type = Union[float, int, IPv4Address, bytearray, None]

@dataclass
class Op_result(ABC):
    val: op_result_type = None

    def __hash__(self): 
        return hash(self.val)

@dataclass
class Float(Op_result):
    val: float

@dataclass
class Int(Op_result):
    val: int

@dataclass
class Ipv4(Op_result):
    val: IPv4Address

@dataclass
class MAC(Op_result):
    val: bytearray

@dataclass
class Empty(Op_result):
    val: None = None


class PacketHeaders:
    def __init__(self, data: dict[str, Op_result] | None=None):
        self.headers: dict[str, 'Op_result'] | None = data if data is not None else {}

    def __iter__(self):
        return iter(self.headers)
    
    def __getitem__(self, key: str):
        if not isinstance(key, str):
            raise TypeError("Packet keys may only be strings.")
        else:
            return self.headers[key]
    
    def __setitem__(self, key: str, val: Op_result):
        if not isinstance(key, str):
            raise TypeError("Packet keys may only be strings.")
        elif not isinstance(val, Op_result):
            raise TypeError("Packet values may only be Op_results.")
        else: 
            self.headers[key] = val
        return self
    
    def items(self) -> ItemsView[str, Op_result]:
        return self.headers.items()
    
    def __hash__(self):
        return hash(frozenset(self.headers.items()))
    
    def __or__(self, other: 'PacketHeaders') -> 'PacketHeaders':
        if not isinstance(other, PacketHeaders):
            raise TypeError("Packets may only be unioned with other Packets.")
        return PacketHeaders(self.headers.__or__(other.headers))
    
    def get(self, key: str, default=None):
        if not isinstance(key, str):
            raise TypeError("Packet keys may only be strings.")
        return self.headers.get(key, default)
    
    def get_mapped_int(self, key: str) -> int:
        if not isinstance(key, str):
            raise TypeError("Packet keys may only be strings.")
        return int_of_op_result(self.headers[key])
    
    def get_mapped_float(self, key: str) -> float:
        if not isinstance(key, str):
            raise TypeError("Packet keys may only be strings.")
        return float_of_op_result(self.headers[key])
    
    def string_of_packet(self) -> str:
        return "".join(f'"{key}" => {string_of_op_result(val)}, ' 
                   for key, val in self.headers)
        
    def packet_of_list(self, packet_list: deque[tuple[str, Op_result]]) -> "PacketHeaders":
        self.headers = {key: val for key, val in list(packet_list)}

    def dump_packet(self, outc: TextIO) -> None:
        print(self.string_of_packet())

    def lookup_int(self, key: str) -> int:
        return int_of_op_result(self.headers[key])

    def lookup_float(self, key: str) -> float:
        return float_of_op_result(self.headers[key])

def string_of_mac(buf: bytearray) -> str:
    return f"{buf[0]: .2f}:{buf[1]: .2f}:{buf[3]: .2f}:\
                {buf[4]: .2f}:{buf[5]: .2f}"

def tcp_flags_to_strings(flags: int) -> str:
    tcp_flags: list[str] = [ 
                            "FIN", "SYN", "RST", "PSH", 
                            "ACK", "URG", "ECE", "CWR", 
                           ]
    def tcp_flags_map() -> Iterator[list[tuple[str, float]]]:
        for i, flag in enumerate(tcp_flags):
            yield (flag, 1 << i)
    return "|".join(flag for flag, val in tcp_flags_map() 
                    if (flags & val) == val)

def int_of_op_result(input: Op_result) -> int:
    match input.kind:
        case Op_result.INT: return input.val
        case _:
            raise TypeError("Trying to extract int from non-int result")

def float_of_op_result(input: Op_result) -> float:
    match input.kind:
        case Op_result.FLOAT: return input.val
        case _:
            raise TypeError("Trying to extract float from non-float result")
        
def string_of_op_result(input: Op_result) -> str:
    match input.kind:
        case Op_result.FLOAT | Op_result.INT:
            return f"{input.val}"
        case Op_result.IPV4:
            return str(IPv4Address(input.val))
        case Op_result.MAC:
            return string_of_mac(input.val)
        case Op_result.Empty:
            "Empty"
        case _:
            raise RuntimeError("Reached unreachable code")
        

File: assisted-translations/functionalist/ts-functionalist/utils-translated.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.opResultKind = void 0;
exports.$π = $π;
exports.$$π = $$π;
exports.tcpFlagsToStrings = tcpFlagsToStrings;
exports.intOfOpResult = intOfOpResult;
exports.floatOfOpResult = floatOfOpResult;
exports.stringOfOpResult = stringOfOpResult;
exports.stringOfPacketHeaders = stringOfPacketHeaders;
exports.packetHeadersOfList = packetHeadersOfList;
exports.dumpPacketHeaders = dumpPacketHeaders;
exports.lookupInt = lookupInt;
exports.lookupFloat = lookupFloat;
var opResultKind;
(function (opResultKind) {
    opResultKind[opResultKind["Float"] = 0] = "Float";
    opResultKind[opResultKind["Int"] = 1] = "Int";
    opResultKind[opResultKind["IPv4"] = 2] = "IPv4";
    opResultKind[opResultKind["MAC"] = 3] = "MAC";
    opResultKind[opResultKind["Empty"] = 4] = "Empty";
})(opResultKind || (exports.opResultKind = opResultKind = {}));
function $π(opCreatorFunc, nextOp) {
    return opCreatorFunc(nextOp);
}
function $$π(opCreatorFunc, nextOp) {
    return opCreatorFunc(nextOp);
}
function tcpFlagsToStrings(flags) {
    let acc = "";
    new Map([
        ["FIN", 1 << 0],
        ["SYN", 1 << 1],
        ["RST", 1 << 2],
        ["PSH", 1 << 3],
        ["ACK", 1 << 4],
        ["URG", 1 << 5],
        ["ECE", 1 << 6],
        ["CWR", 1 << 7],
    ]).forEach((val, key) => {
        if ((flags & val) === val) {
            acc = acc + (acc === "")
                ? `${key}`
                : "|" + `${key}`;
        }
    });
    return acc;
}
function intOfOpResult(input) {
    switch (input.kind) {
        case opResultKind.Int:
            return input.val;
        default:
            throw new TypeError("Trying to extract int from non-int result");
    }
}
function floatOfOpResult(input) {
    switch (input.kind) {
        case opResultKind.Float:
            return input.val;
        default:
            throw new TypeError("Trying to extract float from non-float result");
    }
}
function stringOfOpResult(input) {
    switch (input.kind) {
        case opResultKind.Float:
        case opResultKind.Int:
        case opResultKind.IPv4:
        case opResultKind.MAC:
            return input.val.toString();
        case opResultKind.Empty:
            return "Empty";
    }
}
function stringOfPacketHeaders(input_packet) {
    return ([...input_packet.entries()])
        .reduce((acc, [key, val]) => acc += `${key}" \
                        => ${stringOfOpResult(val)}, `, "");
}
function packetHeadersOfList(header_list) {
    return new Map(header_list);
}
function dumpPacketHeaders(outc, headers) {
    outc.write(`${stringOfPacketHeaders(headers)}`);
}
function lookupInt(key, headers) {
    return intOfOpResult(headers.get(key) ??
        { kind: opResultKind.Empty, val: null });
}
function lookupFloat(key, headers) {
    return floatOfOpResult(headers.get(key) ??
        { kind: opResultKind.Empty, val: null });
}
